### PR TITLE
optional solid color status bar background for widescreen mode

### LINKED ISSUE #651: HUD extenders for widescreen resolutions
Any chance Woof could have [this feature from EDGE classic](https://www.doomworld.com/forum/post/2515544)?
I believe it's a universally accepted fact that GRNROCK is ugly and ruins the look of custom status bars. The solution they found for EDGE classic looks beautiful.

### PR DESCRIPTION
**Pull request by fabiangreffrath**
Fixes #651

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
I have decided to provide a little gradient for the two two rows. What do you say?

woof0001

woof0002

File: src/m_menu.c [hunk 1]
```diff
@@ -3923,6 +3924,9 @@ setup_menu_t gen_settings2[] = { // General Settings screen2
   {"Brightmaps for Textures and Sprites", S_YESNO, m_null, M_X,
    G_Y3 + general_brightmaps*M_SPC, {"brightmaps"}},
 
+  {"Solid Status Bar Background", S_YESNO, m_null, M_X,
```
Comment (fabiangreffrath) [inline src/m_menu.c:3927]: I may need some help with this description from native speakers. @kraflab @mikeday0 Is it clear what is meant or do you have a better wording?

**Comment by fabiangreffrath** [comment]
The diff is best viewed with "Hide whitespace".

File: src/m_menu.c (hunk 1 above)
Comment (fabiangreffrath) [reply src/m_menu.c:3927]: Thank you!

Comment (mikeday0) [reply src/m_menu.c:3927]: I think this is fine as is, but part of me wants something like "Smart Status Bar Background" to give a hint that it's going to analyze the status bar colors and pick something that matches. Cool feature btw!

**Comment by dmslr** [comment]
> I have decided to provide a little gradient for the two two rows. What do you say?

Nice! Can't make it any better.

File: src/m_menu.c (hunk 1 above)
Comment (fabiangreffrath) [reply src/m_menu.c:3927]: "Smart" sounds a bit too bold IMHO, it's still solid colors after all. But thanks for your input!

**Comment by fabiangreffrath** [comment]
I am going to do two more experiments before merging this:

- Reduce the "analyzation depth", i.e. the number of pixels for which the color is averaged on the left and right side of the status bar", to e.g. 12.

- Broaden the gradient to cover line 1 and then lines 2 and 3 instead of only line 1 and line 2.

**Comment by dmslr** [comment]
I tested it a bit with various wads and came to the conclusion that the sides are way too dark compared to STBAR. For example, brown or medium grey status bars make the sides almost black. Maybe it's just me.

**Comment by fabiangreffrath** [comment]
> `I_GetPaletteIndex(pal, r/2, g/2, b/2);`

I think I like this most!

**Comment by fabiangreffrath** [comment]
@dmslr Please feel free to suggest different scaling values for the colors, I am open to suggestions. Though, I have to agree with @rfomin that the undamped average color is too bold.

**Comment by dmslr** [comment]
To be fair, I don't quite understand how all this stuff works. My note about brightness wasn't claimed to be an objective fact and wasn't something that could be taken into consideration. I'm completely satisfied with the result. You and rfomin are the ones who know how it should look like.
Thank you anyway!

**Comment by fabiangreffrath** [comment]
> * Broaden the gradient to cover line 1 and then lines 2 and 3 instead of only line 1 and line 2.

Since we are already in highly subjective ground and I need everybody's opinion to find a reasonable decision anyway, could you guys please try out this little change and tell me if it's a change for better or worse?

```patch
--- a/src/st_stuff.c
+++ b/src/st_stuff.c
@@ -347,7 +347,7 @@ void ST_refreshBackground(boolean force)
         if (st_solidbackground)
         {
           // [FG] calculate average color of the 16px left and right of the status bar
-          const int vstep[][2] = {{0, 1}, {1, 2}, {2, ST_HEIGHT}};
+          const int vstep[][2] = {{0, 1}, {1, 3}, {3, ST_HEIGHT}};
           const int hstep = hires ? (4 * SCREENWIDTH) : SCREENWIDTH;
           const int w = SHORT(sbar->width);
           const int depth = 16;
```

**Comment by rfomin** [comment]
I like the original `{{0, 1}, {1, 2}, {2, ST_HEIGHT}}` the most. I remember trying a similar effect with the sky and also ended up with a thin line.

**Comment by fabiangreffrath** [comment]
> I like the original `{{0, 1}, {1, 2}, {2, ST_HEIGHT}}` the most. I remember trying a similar effect with the sky and also ended up with a thin line.

Thank you! 🙏 So, is every body fine with how it currently looks?

**Comment by mikeday0** [comment]
This is a huge improvement over GRNROCK. Very nicely done!
hell_revealed

plutonia2

tangerine_nightmare

**Pull request merged**